backend/src/main/java/fr/anettehaferkorn/backend/service/MatchingService.java

```diff
@@ -0,0 +1,82 @@
+
+import java.util.List;
+import java.util.Objects;
+
+@Service
+public class MatchingService {
+
+    WineGrapeRepository wineGrapeRepository;
+
+    @Autowired
+    public MatchingService(WineGrapeRepository wineGrapeRepository) {
+        this.wineGrapeRepository = wineGrapeRepository;
+    }
+
+    public List<RecommendationDTO> getMatchingWines(WineQuery wineQuery){
+        List<WineGrape> matchingWineByOccasionAndStyle=matchFirstFilterLayer(wineQuery);
+        List<RecommendationDTO> recommendations= QueryToRecommendationMapper.mapQuerytoRecommendation(matchingWineByOccasionAndStyle);
+        List<RecommendationDTO> optionFilteredRecommendations=matchSecondFilterLayer(wineQuery,recommendations);
+        return cleanMatches(optionFilteredRecommendations);
+
+    }
+    public List<WineGrape> getAllWine(){
+        return wineGrapeRepository.findAll();
+    }
+
+    private List<WineGrape> matchFirstFilterLayer(WineQuery wineQuery){
+        return wineGrapeRepository.findWineQueryDTOByOccasionAndWineStyle
+                (wineQuery.getOccasion(), wineQuery.getWineStyle());
+    }
+
+    @SuppressWarnings("java:S1192")
+    private List<RecommendationDTO> matchSecondFilterLayer(WineQuery wineQuery, List<RecommendationDTO> wineMatches){
+
+        for (RecommendationDTO match: wineMatches) {
+            if(Objects.equals(match.getRegion(), wineQuery.getRegion())||
+                    (Objects.equals(wineQuery.getRegion(),"other"))||
+                    (Objects.equals(wineQuery.getRegion(),"idK"))
+            ){
+                match.setMatchingPoints(match.getMatchingPoints()+1);
+            }
+        }
+        return matchThirdFilterLayer(wineQuery,wineMatches);
+    }
+
+    private List<RecommendationDTO> matchThirdFilterLayer(WineQuery wineQuery, List<RecommendationDTO> wineMatches){
+
+        for (RecommendationDTO match: wineMatches) {
+            if(Objects.equals(match.getAlcohol(), wineQuery.getAlcohol()) ||
+                    (Objects.equals(wineQuery.getRegion(),"other"))||
+                    (Objects.equals(wineQuery.getRegion(),"idK"))){
+                match.setMatchingPoints(match.getMatchingPoints()+1);
+            }
+        }
+        return matchFourthFilterLayer(wineQuery,wineMatches);
+    }
+
+    private List<RecommendationDTO> matchFourthFilterLayer(WineQuery wineQuery, List<RecommendationDTO> wineMatches){
+
+        for (RecommendationDTO match: wineMatches) {
+            if(Objects.equals(match.getTaste(), wineQuery.getTaste()) ||
+                    (Objects.equals(wineQuery.getRegion(),"other"))||
+                    (Objects.equals(wineQuery.getRegion(),"idK"))){
+                match.setMatchingPoints(match.getMatchingPoints()+1);
+            }
+        }
+        return wineMatches;
+    }
+
+    public List<RecommendationDTO>cleanMatches(List<RecommendationDTO> matches){
+            matches.removeIf(match -> match.getMatchingPoints() <2);
+            return matches;
+    }
```

Comment:
I prefer streams, but mabe it's just a matter of taste...
The filters should be named for what they're doing, i.e. filterWinesByRegion
also I wouldn't call the next Filter from each function, but rather return the value and write something like
Wines[] filteredByRegion = fitlterWinesByRegion(query, wines);
Wines[] filterByAlcoholContent = filterWinesByAlcoholContent(filteredByRegion);
...etc.
Thinking about it, they're all not real filters. 
Another possibility to do this would be to actually filter in each of these steps, join all lists together and then select/return only those who occur more than 2 times(or whatever value you choose)

Reply:
The thought behind these three layers was that since they don't have to be set in the frontend (e.g. on I don't mind) they should not exclude entries, but only assign matchingPoints. 
Since all three have to pass through to create the ranking effect, chain them together seemed to make sense.

Reply:
I can see the logic behind it. I just think it's difficult to read.
 I would definitely name the filters for what they're actually filtering for as mentioned above.
Also I think 'cleanMatches' does not state very clearly to the reader what it is actually doing. 
Also I would execute each method in the getMatchingWines method. Like this it will be immediately visible to the reader what is happening in this class
If you would just use filtered lists and then count how often each wine occurs in the combined list you would get around getting/setting matchingPoints in each filter method.